contracts/NodeOperatorRegistry.sol

```diff
@@ -0,0 +1,642 @@
+        _setupRole(REMOVE_OPERATOR_ROLE, msg.sender);
+        _setupRole(EXIT_OPERATOR_ROLE, msg.sender);
+        _setupRole(UPDATE_COMMISSION_RATE_ROLE, msg.sender);
+        _setupRole(UPDATE_STAKE_HEIMDALL_FEES_ROLE, msg.sender);
+    }
+
+    /// @notice Add a new node operator to the system.
+    /// @param _name the node operator name.
+    /// @param _rewardAddress address used for ACL and receive rewards.
+    /// @param _signerPubkey public key used on heimdall len 64 bytes.
+    function addOperator(
+        string memory _name,
+        address _rewardAddress,
+        bytes memory _signerPubkey
+    ) public override whenNotPaused userHasRole(ADD_OPERATOR_ROLE) {
+        require(_signerPubkey.length == 64, "Invalid Public Key");
+        require(_rewardAddress != address(0), "Invalid reward address");
+        require(operatorOwners[_rewardAddress] == 0, "Address already used");
+
+        uint256 operatorId = state.totalNodeOpearator + 1;
+
+        // deploy validator contract.
+        address validatorContract = IValidatorFactory(state.validatorFactory)
+            .create();
+
+        // add the validator.
+        operators[operatorId] = Operator.NodeOperator({
+            status: Operator.NodeOperatorStatus.ACTIVE,
+            name: _name,
+            rewardAddress: _rewardAddress,
+            validatorId: 0,
+            signerPubkey: _signerPubkey,
+            validatorShare: address(0),
+            validatorContract: validatorContract,
+            commissionRate: state.commissionRate
+        });
+
+        // update global state.
+        operatorIds.push(operatorId);
+        state.totalNodeOpearator++;
+        state.totalActiveNodeOpearator++;
+
+        // map user _rewardAddress with the operatorId.
+        operatorOwners[_rewardAddress] = operatorId;
+
+        // emit NewOperator event.
+        emit NewOperator(
+            operatorId,
+            _name,
+            _signerPubkey,
+            Operator.NodeOperatorStatus.ACTIVE
+        );
+    }
+
+    /// @notice Allows to remove an operator from the system.
+    /// @param _operatorId the node operator id.
+    function removeOperator(uint256 _operatorId)
+        public
+        override
+        whenNotPaused
+        userHasRole(REMOVE_OPERATOR_ROLE)
+    {
+        Operator.NodeOperator storage no = operators[_operatorId];
+        require(
+            no.status == Operator.NodeOperatorStatus.EXIT,
+            "Node Operator state not exit"
+        );
+
+        state.totalNodeOpearator--;
+        state.totalExitNodeOpearator--;
+
+        // update the operatorIds array by removing the operator id.
+        for (uint256 idx = 0; idx < operatorIds.length - 1; idx++) {
+            if (_operatorId == operatorIds[idx]) {
+                operatorIds[idx] = operatorIds[operatorIds.length - 1];
+                break;
+            }
+        }
+        delete operatorIds[operatorIds.length - 1];
+        operatorIds.pop();
```

Comment:
```suggestion
        operatorIds.pop();
```
Popping is enough if you want to remove the last element of array. Deleting it actually causes the array at that index to become 0.

Reply:
Sure sure